helix-term/src/commands.rs

```diff
@@ -5750,3 +5755,156 @@ fn replay_macro(cx: &mut Context) {
+                        } else {
+                            from
+                        }
+                    } else {
+                        let to = primary_selection.to();
+                        if range.anchor > to {
+                            range.anchor
+                        } else {
+                            to
+                        }
+                    };
+                    range = Range::new(anchor, range.head);
+                }
+                doc_mut!(cx.editor, &doc).set_selection(view, range.into());
+            }
+        });
+    });
+}
+
+fn jump_to_word(cx: &mut Context, extend: bool) {
+    // Calculate the jump candidates: ranges for any visible words with two or
+    // more characters.
+    let alphabet = &cx.editor.config().jump_label_alphabet;
+    let jump_label_limit = alphabet.len() * alphabet.len();
+    let mut words = Vec::with_capacity(jump_label_limit);
+    let (view, doc) = current!(cx.editor);
+    let text = doc.text().slice(..);
+
+    // This is not necessarily exact if there is virtual text like soft wrap.
+    // It's ok though because the extra jump labels will not be rendered.
+    let start = text.line_to_char(text.char_to_line(view.offset.anchor));
+    let end = text.line_to_char(view.estimate_last_doc_line(doc) + 1);
+
+    let primary_selection = doc.selection(view.id).primary();
+    let cursor = primary_selection.cursor(text);
+    let cursor_fwd = movement::move_prev_word_start(text, Range::point(cursor), 1);
+    let mut cursor_fwd = if cursor_fwd.anchor == cursor {
+        Range::point(cursor_fwd.head)
+    } else {
+        Range::point(cursor)
+    };
+    let mut cursor_rev = cursor_fwd;
+    loop {
+        let mut changed = false;
+        if cursor_fwd.head < end {
+            cursor_fwd = movement::move_next_word_end(text, cursor_fwd, 1);
+            let range = movement::move_prev_word_start(text, cursor_fwd, 1);
+            // The cursor is on a word longer than 2 characters.
+            if RopeGraphemes::new(text.slice(range.head..))
+                .take(2)
+                .all(|g| g.chars().all(char_is_word))
+            {
+                words.push(range.flip());
+            }
+            changed = true;
+            if words.len() == jump_label_limit {
+                break;
+            }
+        }
+        if cursor_rev.head > start {
+            cursor_rev = movement::move_prev_word_start(text, cursor_rev, 1);
+            // The cursor is on a word longer than 2 characters.
+            if RopeGraphemes::new(text.slice(cursor_rev.head..))
+                .take(2)
+                .all(|g| g.chars().all(char_is_word))
+            {
+                words.push(movement::move_next_word_end(text, cursor_rev, 1).flip());
+            }
+            changed = true;
+            if words.len() == jump_label_limit {
+                break;
+            }
+        }
+        if !changed {
+            break;
+        }
+    }
```

Comment:
**thought**: this could reduce the nesting and (subjectively) make it easier to read:
```suggestion
    let mut cursor_rev = cursor_fwd;

    // collect words after the cursor
    while cursor_fwd.head < end && words.len() < jump_label_limit {
        cursor_fwd = movement::move_next_word_end(text, cursor_fwd, 1);
        let range = movement::move_prev_word_start(text, cursor_fwd, 1);
        // The cursor is on a word longer than 2 characters.
        if RopeGraphemes::new(text.slice(range.head..))
            .take(2)
            .all(|g| g.chars().all(char_is_word))
        {
            words.push(range.flip());
        }
    }

    // collect words before the cursor
    while cursor_rev.head > start && words.len() < jump_label_limit {
        cursor_rev = movement::move_prev_word_start(text, cursor_rev, 1);
        // The cursor is on a word longer than 2 characters.
        if RopeGraphemes::new(text.slice(cursor_rev.head..))
            .take(2)
            .all(|g| g.chars().all(char_is_word))
        {
            words.push(movement::move_next_word_end(text, cursor_rev, 1).flip());
        }
    }
```
I hope I didn't miss something in regards to the logic.

Reply:
that is not equivalent tough. I purposefully iterate forward and reverse in lockstep so that labels are assigned in order of the textual distance.

Reply:
Ah, gotcha. I see now. Thank you.